File: test/aidbox_sdk/snapshots/python/hl7_fhir_r4_core/AuditEvent.py
from typing import Optional, List
from dataclasses import dataclass, field
from base import Period
from base import CodeableConcept
from base import Coding
from base import DomainResource
from base import Reference
from base import BackboneElement

@dataclass(kw_only=True)
class AuditEventSource(BackboneElement):
    site: Optional[str] = None
    type: Optional[List[Coding]] = field(default_factory=list)
    observer: Reference

@dataclass(kw_only=True)
class AuditEventAgentNetwork(BackboneElement):
    type: Optional[str] = None
    address: Optional[str] = None

@dataclass(kw_only=True)
class AuditEventAgent(BackboneElement):
    role: Optional[List[CodeableConcept]] = field(default_factory=list)
    requestor: bool
    who: Optional[Reference] = None
    alt_id: Optional[str] = None
    name: Optional[str] = None
    type: Optional[CodeableConcept] = None
    policy: Optional[List[str]] = field(default_factory=list)
    purpose_of_use: Optional[List[CodeableConcept]] = field(default_factory=list)
    network: Optional[AuditEventAgentNetwork] = None
    location: Optional[Reference] = None
    media: Optional[Coding] = None

@dataclass(kw_only=True)
class AuditEventEntityDetail(BackboneElement):
    type: str
    value_string: Optional[str] = None
    value_base64_binary: Optional[str] = None

@dataclass(kw_only=True)
class AuditEventEntity(BackboneElement):
    role: Optional[Coding] = None
    description: Optional[str] = None
    name: Optional[str] = None
    type: Optional[Coding] = None
    lifecycle: Optional[Coding] = None
    query: Optional[str] = None
    security_label: Optional[List[Coding]] = field(default_factory=list)
    what: Optional[Reference] = None
    detail: Optional[List[AuditEventEntityDetail]] = field(default_factory=list)

@dataclass(kw_only=True)
class AuditEvent(DomainResource):
    outcome_desc: Optional[str] = None
    type: Coding
    outcome: Optional[str] = None
    source: AuditEventSource
    recorded: str
    agent: list[AuditEventAgent] = field(default_factory=list)
    purpose_of_event: Optional[List[CodeableConcept]] = field(default_factory=list)
    action: Optional[str] = None
    period: Optional[Period] = None
    entity: Optional[List[AuditEventEntity]] = field(default_factory=list)
    subtype: Optional[List[Coding]] = field(default_factory=list)

File: test/aidbox_sdk/snapshots/dotnet/hl7-fhir-r4-core/AuditEvent.cs
using Aidbox.FHIR.Base;
using Aidbox.FHIR.Utils;

namespace Aidbox.FHIR.R4.Core;

public class AuditEvent : DomainResource, IResource
{
    public string? OutcomeDesc { get; set; }
    public required Base.Coding Type { get; set; }
    public string? Outcome { get; set; }
    public required AuditEvent_Source Source { get; set; }
    public required string Recorded { get; set; }
    public required AuditEvent_Agent[] Agent { get; set; }
    public Base.CodeableConcept[]? PurposeOfEvent { get; set; }
    public string? Action { get; set; }
    public Base.Period? Period { get; set; }
    public AuditEvent_Entity[]? Entity { get; set; }
    public Base.Coding[]? Subtype { get; set; }

    public class AuditEvent_Source : BackboneElement
    {
        public string? Site { get; set; }
        public Base.Coding[]? Type { get; set; }
        public required Base.ResourceReference Observer { get; set; }
    }

    public class AuditEvent_Agent_Network : BackboneElement
    {
        public string? Type { get; set; }
        public string? Address { get; set; }
    }

    public class AuditEvent_Agent : BackboneElement
    {
        public Base.CodeableConcept[]? Role { get; set; }
        public required bool Requestor { get; set; }
        public Base.ResourceReference? Who { get; set; }
        public string? AltId { get; set; }
        public string? Name { get; set; }
        public Base.CodeableConcept? Type { get; set; }
        public string[]? Policy { get; set; }
        public Base.CodeableConcept[]? PurposeOfUse { get; set; }
        public AuditEvent_Agent_Network? Network { get; set; }
        public Base.ResourceReference? Location { get; set; }
        public Base.Coding? Media { get; set; }
    }

    public class AuditEvent_Entity_Detail : BackboneElement
    {
        public required string Type { get; set; }
        public object? Value    
        {
            get
            {
                if (ValueString is not null)
                {
                    return ValueString;
                }
        
                if (ValueBase64Binary is not null)
                {
                    return ValueBase64Binary;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(string))
                {
                    ValueString = (string)value;return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    ValueBase64Binary = (string)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public string? ValueString { get; set; }
        public string? ValueBase64Binary { get; set; }
    }

    public class AuditEvent_Entity : BackboneElement
    {
        public Base.Coding? Role { get; set; }
        public string? Description { get; set; }
        public string? Name { get; set; }
        public Base.Coding? Type { get; set; }
        public Base.Coding? Lifecycle { get; set; }
        public string? Query { get; set; }
        public Base.Coding[]? SecurityLabel { get; set; }
        public Base.ResourceReference? What { get; set; }
        public AuditEvent_Entity_Detail[]? Detail { get; set; }
    }
}